### PR TITLE
Modify kapi_im module to use kapi_definition module

### PR DESCRIPTION
**Pull request by harenson**

### CODE REVIEW

**Comment by harenson** [comment]
@lazedo sure thing! thanks for reviewing 👍

File: core/kazoo_amqp/src/api/kapi_im.erl [hunk 1]
```diff
@@ -11,137 +11,210 @@
...
+              ,{fun kapi_definition:set_optional_headers/2, [<<"Account-ID">>
+                                                            ,<<"Charges">>
+                                                            ,<<"Custom-SIP-Headers">>
+                                                            ,<<"Custom-Vars">>
+                                                            ,<<"Request">>
+                                                            ,<<"Route-ID">>
+                                                            ,<<"Route-Type">>
+                                                            ,<<"System-ID">>
+                                                            ,<<"Transaction-ID">>
+                                                            ]}
```
Comment (lazedo) [inline core/kazoo_amqp/src/api/kapi_im.erl]: why are you defining the headers here and not reuse the macros instead ?

File: core/kazoo_amqp/src/api/kapi_im.erl [hunk 2]
```diff
@@ -11,137 +11,210 @@
...
+    outbound_definition().
+
+-spec inbound_definition() -> kapi_definition:api().
+inbound_definition() ->
+    EventName = <<"inbound">>,
+    Category = ?EVENT_CATEGORIES,
+    Setters = [{fun kapi_definition:set_name/2, EventName}
+              ,{fun kapi_definition:set_friendly_name/2, <<"Inbound">>}
+              ,{fun kapi_definition:set_description/2, <<"SMS MMS Inbound">>}
+              ,{fun kapi_definition:set_category/2, Category}
```
Comment (lazedo) [inline core/kazoo_amqp/src/api/kapi_im.erl]: you're setting a list of binaries in a api_ne_binary

File: core/kazoo_amqp/src/api/kapi_im.erl [hunk 3]
```diff
@@ -11,137 +11,210 @@
...
+                                                            ,<<"Originator-Flags">>
+                                                            ,<<"Originator-Properties">>
+                                                            ,<<"Request">>
+                                                            ,<<"Route-ID">>
+                                                            ,<<"Route-Type">>
+                                                            ,<<"System-ID">>
+                                                            ,<<"Target-Flags">>
+                                                            ,<<"Target-Properties">>
+                                                            ,<<"Transaction-ID">>
+                                                            ]}
```
Comment (lazedo) [inline core/kazoo_amqp/src/api/kapi_im.erl]: same here for the macro comment

File: core/kazoo_amqp/src/api/kapi_im.erl [hunk 4]
```diff
@@ -11,137 +11,210 @@
...
+              ,{fun kapi_definition:set_values/2
+               ,[{<<"Route-Type">>, [<<"onnet">>, <<"offnet">>]}
+                 | kapi_definition:event_type_headers(Category, EventName)
```
Comment (lazedo) [inline core/kazoo_amqp/src/api/kapi_im.erl]: `Category` from macro is a list of binaries, does this work ?

File: core/kazoo_amqp/src/api/kapi_im.erl [hunk 5]
```diff
@@ -11,137 +11,210 @@
...
+               }
+              ,{fun kapi_definition:set_types/2
+               ,[{<<"Body">>, fun is_binary/1}
+                ,{<<"Custom-Vars">>, fun kz_json:is_json_object/1}
+                ,{<<"Endpoints">>, fun is_list/1}
+                ,{<<"From">>, fun is_binary/1}
+                ,{<<"Message-ID">>, fun is_binary/1}
+                ,{<<"Request">>, fun is_binary/1}
+                ,{<<"System-ID">>, fun is_binary/1}
+                ,{<<"To">>, fun is_binary/1}
```
Comment (lazedo) [inline core/kazoo_amqp/src/api/kapi_im.erl]: more macro

File: core/kazoo_amqp/src/api/kapi_im.erl [hunk 6]
```diff
@@ -11,137 +11,210 @@
...
+
+-spec publish_inbound(kz_term:api_terms(), kz_term:proplist()) -> 'ok'.
+publish_inbound(Req, AMQPOptions) ->
+    Definition = inbound_definition(),
+    {'ok', Payload} = kz_api:prepare_api_payload(Req
+                                                ,kapi_definition:values(Definition)
+                                                ,kapi_definition:build_fun(Definition)
+                                                ),
+    kz_amqp_util:basic_publish(exchange_id(Req)
+                              ,(kapi_definition:binding(Definition))(Req)
```
Comment (lazedo) [inline core/kazoo_amqp/src/api/kapi_im.erl]: this code is a bit confusing. i know it return a `fun` but still confusing. maybe add `kapi_definition:binding/2`   that takes definition and api_terms and returns the routing key

File: core/kazoo_amqp/src/api/kapi_im.erl [hunk 7]
```diff
@@ -11,137 +11,210 @@
...
+
+-spec publish_outbound(kz_term:api_terms(), kz_term:proplist()) -> 'ok'.
+publish_outbound(Req, AMQPOptions) ->
+    Definition = outbound_definition(),
+    {'ok', Payload} = kz_api:prepare_api_payload(Req
+                                                ,kapi_definition:values(Definition)
+                                                ,kapi_definition:build_fun(Definition)
+                                                ),
+    kz_amqp_util:basic_publish(exchange_id(Req)
+                              ,(kapi_definition:binding(Definition))(Req)
```
Comment (lazedo) [inline core/kazoo_amqp/src/api/kapi_im.erl]: same here for the kapi_definition

File: core/kazoo_amqp/src/api/kapi_im.erl [hunk 8]
```diff
@@ -153,20 +226,32 @@ bind_q(Queue, Props) ->
...
+                                        )
          || Category <- bind_categories(Props)
         ],
     bind_q(Queue, Props, Restrict);
 bind_q(Queue, Props, ['outbound'|Restrict]) ->
-    _ = [kz_amqp_util:bind_q_to_exchange(Queue, ?BIND_OUTBOUND_RK(Category, Props), bind_exchange_id(Props))
+    _ = [kz_amqp_util:bind_q_to_exchange(Queue
+                                        ,bind_outbound_routing_key(Category, Props)
+                                        ,bind_exchange_id(Props)
+                                        )
```
Comment (lazedo) [inline core/kazoo_amqp/src/api/kapi_im.erl]: more macro

File: core/kazoo_amqp/src/api/kapi_im.erl [hunk 9]
```diff
@@ -178,20 +263,32 @@ unbind_q(Queue, Props) ->
...
+                                            )
          || Category <- bind_categories(Props)
         ],
     unbind_q(Queue, Props, Restrict);
 unbind_q(Queue, Props, ['outbound'|Restrict]) ->
-    _ = [kz_amqp_util:unbind_q_from_exchange(Queue, ?BIND_OUTBOUND_RK(Category, Props), bind_exchange_id(Props))
+    _ = [kz_amqp_util:unbind_q_from_exchange(Queue
+                                            ,bind_outbound_routing_key(Category, Props)
+                                            ,bind_exchange_id(Props)
+                                            )
```
Comment (lazedo) [inline core/kazoo_amqp/src/api/kapi_im.erl]: more macro

File: core/kazoo_amqp/src/api/kapi_im.erl [hunk 10]
```diff
@@ -215,7 +312,7 @@ bind_route_id(Props) ->
 
 -spec bind_route_type(kz_term:api_terms()) -> kz_term:ne_binary().
 bind_route_type(Props) ->
-    ?LOWER(props:get_value('route_type', Props, <<"*">>)).
+    to_lower(props:get_value('route_type', Props, <<"*">>)).
```
Comment (lazedo) [inline core/kazoo_amqp/src/api/kapi_im.erl]: why this change `?LOWER` to `to_lower` ?

**Comment by lazedo** [comment]
@harenson reopen after the breakening

**Comment by harenson** [comment]
@lazedo Reopened: https://github.com/2600hz/kazoo-core/pull/30

File: core/kazoo_amqp/src/api/kapi_im.erl (hunk 1 above)
Comment (harenson) [reply core/kazoo_amqp/src/api/kapi_im.erl:91]: Because only the definition function should know about those, if we leave them on the macros then they could be used by any other function instead of using `kapi_definition` module getter functions which is the intention with these changes.

File: core/kazoo_amqp/src/api/kapi_im.erl (hunk 2 above)
Comment (harenson) [reply core/kazoo_amqp/src/api/kapi_im.erl:72]: Fixed on the new PR.

File: core/kazoo_amqp/src/api/kapi_im.erl (hunk 6 above)
Comment (harenson) [reply core/kazoo_amqp/src/api/kapi_im.erl:183]: It is not possible because every module uses a different number of values for constructing routing keys.

File: core/kazoo_amqp/src/api/kapi_im.erl (hunk 1 above)
Comment (lazedo) [reply core/kazoo_amqp/src/api/kapi_im.erl:91]: that doesn't make sense. i prefer to have the macros
